File: exercise/C#/day02/Games.Tests/FizzBuzzTests.cs
using System.Collections.Generic;
using System.Linq;
using FluentAssertions;
using FluentAssertions.LanguageExt;
using FsCheck;
using FsCheck.Xunit;
using LanguageExt;
using Xunit;

namespace Games.Tests
{
    public class FizzBuzzTests
    {
        private static readonly string[] FizzBuzzStrings = ["Fizz", "Buzz", "FizzBuzz"];

        private static readonly Dictionary<int, string> Mapping = new()
        {
            { 3, "Fizz" },
            { 5, "Buzz" }
        };

        private static FizzBuzz FizzBuzzInstance => new(Mapping);

        private static FizzBuzz WhizzBangInstance => new(new Dictionary<int, string>
        {
            { 11, "Bang" },
            { 7, "Whizz" }
        });

        public static IEnumerable<object[]> FizzBuzzData()
        {
            yield return [1, "1"];
            yield return [67, "67"];
            yield return [82, "82"];
            yield return [3, "Fizz"];
            yield return [66, "Fizz"];
            yield return [99, "Fizz"];
            yield return [5, "Buzz"];
            yield return [50, "Buzz"];
            yield return [85, "Buzz"];
            yield return [15, "FizzBuzz"];
            yield return [30, "FizzBuzz"];
            yield return [45, "FizzBuzz"];
        }


        [Theory]
        [MemberData(nameof(FizzBuzzData))]
        public void Returns_Number_Representation(int input, string expectedResult)
        {
            FizzBuzzInstance.Convert(input)
                .Should()
                .BeSome(x => x.Should().Be(expectedResult));
        }

        [Theory]
        [MemberData(nameof(WhizzBangData))]
        public void Returns_WhizzBang_Number_Representation(int input, string expectedResult)
        {
            WhizzBangInstance.Convert(input)
                .Should()
                .BeSome(x => x.Should().Be(expectedResult));
        }

        public static IEnumerable<object[]> WhizzBangData()
        {
            yield return [7, "Whizz"];
            yield return [8, "8"];
            yield return [11, "Bang"];
            yield return [14, "Whizz"];
            yield return [22, "Bang"];
            yield return [77, "WhizzBang"];
            yield return [88, "Bang"];
            yield return [55, "Bang"];
        }

        [Property]
        public Property Parse_Return_Valid_String_For_Numbers_Between_1_And_100()
            => Prop.ForAll(
                ValidInput(),
                IsConvertValid
            );

        private static Arbitrary<int> ValidInput()
            => Gen.Choose(FizzBuzz.Min, FizzBuzz.Max).ToArbitrary();

        private static bool IsConvertValid(int x)
            => FizzBuzzInstance.Convert(x).Exists(s => ValidStringsFor(x).Contains(s));

        private static IEnumerable<string> ValidStringsFor(int x)
            => FizzBuzzStrings.Append(x.ToString());

        [Property]
        public Property ParseFailForNumbersOutOfRange()
            => Prop.ForAll(
                InvalidInput(),
                x => FizzBuzzInstance.Convert(x).IsNone
            );

        private static Arbitrary<int> InvalidInput()
            => Gen.Choose(-10_000, 10_000)
                .ToArbitrary()
                .Filter(x => x is < FizzBuzz.Min or > FizzBuzz.Max);
    }
}

File: exercise/java/day02/src/test/java/games/FizzBuzzTests.java
package games;

import io.vavr.collection.LinkedHashMap;
import io.vavr.collection.Seq;
import io.vavr.test.Arbitrary;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

import static games.FizzBuzz.MAX;
import static games.FizzBuzz.MIN;
import static io.vavr.API.List;
import static io.vavr.API.Some;
import static io.vavr.test.Arbitrary.integer;
import static io.vavr.test.Property.def;
import static org.assertj.core.api.Assertions.assertThat;

class FizzBuzzTests {
    private static final Seq<String> fizzBuzzStrings = List("Fizz", "Buzz", "FizzBuzz");

    private static final FizzBuzz fizzBuzz = new FizzBuzz(LinkedHashMap.of(
            15, "FizzBuzz",
            3, "Fizz",
            5, "Buzz"
    ));

    private static final FizzBuzz whizzBang = new FizzBuzz(LinkedHashMap.of(
            7, "Whizz",
            11, "Bang"
    ));

    public static Stream<Arguments> fizzBuzzValidInputs() {
        return Stream.of(
                Arguments.of(1, "1"),
                Arguments.of(67, "67"),
                Arguments.of(82, "82"),
                Arguments.of(3, "Fizz"),
                Arguments.of(66, "Fizz"),
                Arguments.of(99, "Fizz"),
                Arguments.of(5, "Buzz"),
                Arguments.of(50, "Buzz"),
                Arguments.of(85, "Buzz"),
                Arguments.of(15, "FizzBuzz"),
                Arguments.of(30, "FizzBuzz"),
                Arguments.of(45, "FizzBuzz")
        );
    }

    public static Stream<Arguments> whizzBangValidInputs() {
        return Stream.of(
                Arguments.of(7, "Whizz"        ),
                Arguments.of(8, "8"     ),
                Arguments.of(11, "Bang" ),
                Arguments.of(14, "Whizz" ),
                Arguments.of(22, "Bang"),
                Arguments.of(77, "Whizz"         ),
                Arguments.of(88, "Bang"),
                Arguments.of(55, "Bang")
        );
    }

    @ParameterizedTest
    @MethodSource("fizzBuzzValidInputs")
    void parse_successfully_numbers_between_1_and_100_samples_using_fizzBuzz(int input, String expectedResult) {
        assertThat(fizzBuzz.convert(input))
                .isEqualTo(Some(expectedResult));
    }

    @ParameterizedTest
    @MethodSource("whizzBangValidInputs")
    void parse_successfully_numbers_between_1_and_100_samples_using_whizzBang(int input, String expectedResult) {
        assertThat(whizzBang.convert(input))
                .isEqualTo(Some(expectedResult));
    }


    @Test
    void parse_return_valid_string_for_numbers_between_1_and_100() {
        def("Some(validString) for numbers in [1; 100]")
                .forAll(validInput())
                .suchThat(this::isConvertValid)
                .check()
                .assertIsSatisfied();
    }

    @Test
    void parse_fail_for_numbers_out_of_range() {
        def("None for numbers out of range")
                .forAll(invalidInput())
                .suchThat(x -> fizzBuzz.convert(x).isEmpty())
                .check()
                .assertIsSatisfied();
    }

    private boolean isConvertValid(Integer x) {
        return fizzBuzz.convert(x)
                .exists(s -> validStringsFor(x).contains(s));
    }

    private static Arbitrary<Integer> validInput() {
        return integer().filter(x -> x >= MIN && x <= MAX);
    }

    private static Seq<String> validStringsFor(Integer x) {
        return fizzBuzzStrings.append(x.toString());
    }

    private static Arbitrary<Integer> invalidInput() {
        return integer().filter(x -> x < MIN || x > MAX);
    }
}